File: go/route.go
package main

import (
	"encoding/binary"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"time"

	"github.com/doraemonkeys/WindSend/language"
	"github.com/sirupsen/logrus"
	"golang.design/x/clipboard"
)

const TimeFormat = "2006-01-02 15:04:05"

const MaxTimeDiff float64 = 300

const (
	// 服务器内部错误
	ErrorInternal = "internal error"
	// 无效的验证数据
	ErrorInvalidAuthData = "invalid auth data"
	// 过期的验证数据
	ErrorExpiredAuthData = "expired auth data"
	// 剪切板数据过大
	ErrorClipboardDataTooLarge = "clipboard data too large"
	// 损坏的数据
	ErrorInvalidData = "invalid data"
	// 不完整的数据
	ErrorIncompleteData = "incomplete data"
)

const (
	pasteTextAction = "pasteText"
	pasteFileAction = "pasteFile"
	copyAction      = "copy"
	pingAction      = "ping"
	downloadAction  = "download"
	matchAction     = "match"
	syncTextAction  = "syncText"
	webIp           = "web"
)

type headInfo struct {
	Action     string       `json:"action"`
	DeviceName string       `json:"deviceName"`
	TimeIp     string       `json:"timeIp"`
	FileID     uint32       `json:"fileID"`
	FileSize   int64        `json:"fileSize"`
	Path       string       `json:"path"`
	UploadType pathInfoType `json:"uploadType"`
	// Name       string `json:"name"`
	Start   int64 `json:"start"`
	End     int64 `json:"end"`
	DataLen int64 `json:"dataLen"`
	// 操作ID
	OpID uint32 `json:"opID"`
	// 此次操作想要上传的文件数量
	FilesCountInThisOp int `json:"filesCountInThisOp"`
	// Msg      string `json:"msg"`
}

type RespHead struct {
	Code int `json:"code"`
	// TimeIp string   `json:"timeIp"`
	Msg string `json:"msg"`
	// 客户端copy时返回的数据类型(text, image, file)
	DataType string `json:"dataType"`
	// 如果body有数据，返回数据的长度
	DataLen int64 `json:"dataLen"`
	// Paths   []pathInfo `json:"paths"`
}

type pathInfo struct {
	// dir or file
	Type     pathInfoType `json:"type"`
	Path     string       `json:"path"`
	SavePath string       `json:"savePath"`
	Size     int64        `json:"size"`
}

type pathInfoType = string

const (
	pathInfoTypeDir  pathInfoType = "dir"
	pathInfoTypeFile pathInfoType = "file"
)

type MatchActionResp struct {
	DeviceName   string `json:"deviceName"`
	SecretKeyHex string `json:"secretKeyHex"`
}

const (
	DataTypeText      = "text"
	DataTypeClipImage = "clip-image"
	DataTypeFilePaths = "files"
	DataTypeBinary    = "binary"
)

var panicWriter = NewLazyFileWriter("panic.log")

func mainProcess(conn net.Conn) {
	defer func() {
		if err := recover(); err != nil {
			logrus.Error("panic:", err)
			_, _ = panicWriter.Write([]byte(fmt.Sprintf("%v\n", err)))
			var stackBuf = make([]byte, 1024*1024)
			n := runtime.Stack(stackBuf, false)
			_, _ = panicWriter.Write(stackBuf[:n])
		}
	}()
	logrus.Info("remote addr:", conn.RemoteAddr().String())
	defer conn.Close()

	for {
		head, ok := commonAuth(conn)
		if !ok {
			return
		}

		switch head.Action {
		case pingAction:
			pingHandler(conn, head)
			return
		case pasteTextAction:
			pasteTextHandler(conn, head)
			return
		case pasteFileAction:
			ok = pasteFileHandler(conn, head)
		case copyAction:
			copyHandler(conn)
			return
		case downloadAction:
			ok = downloadHandler(conn, head)
		case matchAction:
			matchHandler(conn)
			return
		case syncTextAction:
			syncTextHandler(conn, head)
		default:
			respCommonError(conn, "unknown action:"+head.Action)
			logrus.Error("unknown action:", head.Action)
			return
		}
		if !ok {
			return
		}
	}
}

func pasteTextHandler(conn net.Conn, head headInfo) {
	var bodyBuf = make([]byte, head.DataLen)
	_, err := io.ReadFull(conn, bodyBuf)
	if err != nil {
		// logrus.Error("read body error: ", err)
		logrus.Errorf("read body error: %v, dataLen:%d, bodyBuf:%s\n", err, head.DataLen, string(bodyBuf))
		logrus.Info("head:", head)
		respCommonError(conn, ErrorIncompleteData+": "+err.Error())
		return
	}
	clipboard.Write(clipboard.FmtText, bodyBuf)

	var completionSignal = make(chan struct{})

	go func() {
		// time.Sleep(time.Millisecond * 100)
		_ = sendMsg(conn, "粘贴成功")
		completionSignal <- struct{}{}
	}()

	contentRune := []rune(string(bodyBuf))
	showLen := 60
	if len(contentRune) >= showLen {
		Inform(string(contentRune[:showLen])+"...", head.DeviceName)
	} else {
		Inform(string(contentRune), head.DeviceName)
	}
	<-completionSignal
}

func syncTextHandler(conn net.Conn, head headInfo) {
	var curClipboardText = make([]byte, 0)
	if clipboarDataType == clipboardWatchDataTypeText {
		curClipboardText = clipboardWatchData
	}
	if head.DataLen > 0 {
		var bodyBuf = make([]byte, head.DataLen)
		_, err := io.ReadFull(conn, bodyBuf)
		if err != nil {
			// logrus.Error("read body error: ", err)
			logrus.Errorf("read body error: %v, dataLen:%d, bodyBuf:%s\n", err, head.DataLen, string(bodyBuf))
			logrus.Info("head:", head)
			respCommonError(conn, ErrorIncompleteData+": "+err.Error())
			return
		}
		clipboard.Write(clipboard.FmtText, bodyBuf)
	}
	var _ = sendMsgWithBody(conn, "SyncSuccess", DataTypeText, curClipboardText)
}

func pingHandler(conn net.Conn, head headInfo) {
	var bodyBuf = make([]byte, head.DataLen)
	_, err := io.ReadFull(conn, bodyBuf)
	if err != nil {
		logrus.Error("read body error: ", err)
		return
	}
	decryptedBody, err := crypter.Decrypt(bodyBuf)
	if err != nil {
		logrus.Error("decrypt body error: ", err)
		return
	}
	if string(decryptedBody) != "ping" {
		logrus.Error("invalid ping data: ", string(decryptedBody))
		respCommonError(conn, ErrorInvalidData)
		return
	}
	resp := "pong"
	encryptedResp, err := crypter.Encrypt([]byte(resp))
	if err != nil {
		logrus.Error("encrypt body error: ", err)
		return
	}
	// encryptedResp = []byte("pong")
	_ = sendMsgWithBody(conn, language.Translate(language.VerifySuccess), DataTypeText, encryptedResp)
}

func readHead(conn net.Conn) (headInfo, error) {
	var headBuf = make([]byte, 4)
	var head headInfo

	// 读取json长度
	var headLen int32
	if _, err := io.ReadFull(conn, headBuf[:4]); err != nil {
		return head, err
	}
	headLen = int32(binary.LittleEndian.Uint32(headBuf[:4]))
	// 头部不能超过10KB，以防止恶意攻击导致内存溢出
	const maxHeadLen = 1024 * 10
	if headLen > maxHeadLen {
		return head, fmt.Errorf("head len too large: %d", headLen)
	}
	headBuf = make([]byte, headLen)
	// 读取json
	if _, err := io.ReadFull(conn, headBuf[:headLen]); err != nil {
		logrus.Error("read head failed, err:", err)
		return head, err
	}

	if err := json.Unmarshal(headBuf[:headLen], &head); err != nil {
		logrus.Error("json unmarshal failed, err:", err)
		return head, err
	}
	return head, nil
}

func commonAuth(conn net.Conn) (headInfo, bool) {
	const unauthorizedCode = 401
	logrus.Debugln("commonAuth remote addr:", conn.RemoteAddr().String())

	head, err := readHead(conn)
	if err != nil {
		if errors.Is(err, io.EOF) {
			logrus.Infof("client closed, ip:%s\n", conn.RemoteAddr().String())
		} else {
			respError(conn, unauthorizedCode, err.Error())
			logrus.Error("read head failed, err:", err)
		}
		return head, false
	}
	logrus.Debugf("head: %+v", head)
	if head.Action == matchAction {
		if allowSearch {
			return head, true
		}
		logrus.Infof("search not allowed, deviceName:%s,ip:%s\n", head.DeviceName, conn.RemoteAddr().String())
		respError(conn, unauthorizedCode, "search not allowed")
		return head, false
	}
	if head.TimeIp == "" {
		respError(conn, unauthorizedCode, "time-ip is empty")
		return head, false
	}
	timeAndIPBytes, err := hex.DecodeString(head.TimeIp)
	if err != nil {
		respError(conn, unauthorizedCode, err.Error())
		return head, false
	}
	decrypted, err := crypter.Decrypt(timeAndIPBytes)
	if err != nil {
		respError(conn, unauthorizedCode, err.Error())
		return head, false
	}
	// 2006-01-02 15:04:05
	timeAndIPStr := string(decrypted)
	timeLen := len(TimeFormat)
	if len(timeAndIPStr) < timeLen {
		respError(conn, unauthorizedCode, "time-ip is too short")
		return head, false
	}
	timeStr := timeAndIPStr[:timeLen]
	ip := timeAndIPStr[timeLen+1:]
	t, err := time.Parse(TimeFormat, timeStr)
	if err != nil {
		respError(conn, unauthorizedCode, err.Error())
		return head, false
	}
	if time.Since(t).Seconds() > MaxTimeDiff {
		logrus.Info("time expired: ", t.String())
		respError(conn, unauthorizedCode, "time expired: "+t.String())
		return head, false
	}

	var myip = conn.LocalAddr().String()[0:strings.LastIndex(conn.LocalAddr().String(), ":")]
	if myip[0] == '[' && myip[len(myip)-1] == ']' {
		myip = myip[1 : len(myip)-1] //去掉[]
	}
	if strings.Contains(myip, "%") {
		// ipv6, 去掉%后的作用域id
		myip = myip[0:strings.Index(myip, "%")]
	}
	if strings.Contains(ip, "%") {
		ip = ip[:strings.Index(ip, "%")]
	}
	if ip != myip {
		if len(GloballCnf.ExternalIPs) > 0 {
			for _, v := range GloballCnf.ExternalIPs {
				if v == ip {
					return head, true
				}
			}
		}
		logrus.Infoln("ip not match, received ip:", ip, "my ip:", myip)
		respError(conn, unauthorizedCode, fmt.Sprintf("ip not match: %s != %s", ip, myip))
		return head, false
	}
	return head, true
}

func matchHandler(conn net.Conn) {
	resp := MatchActionResp{
		DeviceName:   GetDeviceName(),
		SecretKeyHex: GloballCnf.SecretKeyHex,
	}
	respBuf, err := json.Marshal(resp)
	if err != nil {
		logrus.Error("json marshal failed, err:", err)
		return
	}
	_ = sendMsg(conn, string(respBuf))
	closeAllowSearchCH <- struct{}{}
}

func copyHandler(conn net.Conn) {

	// 用户选择的文件
	if len(SelectedFiles) != 0 {
		err := sendFiles(conn)
		if err != nil {
			logrus.Error("send files error: ", err)
		} else {
			clearFilesCH <- struct{}{}
		}
		return
	}

	// 文件剪切板
	files, err := clipboardS.Files()
	if err == nil && len(files) != 0 {
		SelectedFiles = files
		err := sendFiles(conn)
		if err != nil {
			logrus.Error("send files error: ", err)
		} else {
			SelectedFiles = nil
			err = clipboardS.Clear()
			if err != nil {
				logrus.Error("clear clipboard error: ", err)
			}
		}
		return
	}

	// 空剪切板
	if clipboarDataType == clipboardWatchDataEmpty {
		respCommonError(conn, language.Translate(language.ClipboardIsEmpty))
		return
	}

	// 文本剪切板
	if clipboarDataType == clipboardWatchDataTypeText {
		sendText(conn)
		return
	}

	// 图片剪切板
	if clipboarDataType == clipboardWatchDataTypeImage {
		sendImage(conn)
		return
	}
}

func sendFiles(conn net.Conn) error {
	// resp.Paths = SelectedFiles
	var respPaths = make([]pathInfo, 0, len(SelectedFiles))
	for _, path1 := range SelectedFiles {
		path1 = strings.ReplaceAll(path1, "\\", "/")
		fInfo, err := os.Stat(path1)
		if err != nil {
			logrus.Error("stat file error: ", err)
			respCommonError(conn, err.Error())
			return err
		}
		var pi pathInfo
		pi.Path = path1
		if !fInfo.IsDir() {
			pi.Type = pathInfoTypeFile
			pi.Size = fInfo.Size()
			respPaths = append(respPaths, pi)
			continue
		} else {
			pi.Type = pathInfoTypeDir
			pi.SavePath = filepath.Base(path1)
			respPaths = append(respPaths, pi)
		}
		// 遍历目录
		err = filepath.Walk(path1, func(path2 string, info os.FileInfo, err error) error {
			if err != nil {
				logrus.Error("walk file error: ", err)
				return err
			}
			path2 = strings.ReplaceAll(path2, "\\", "/")
			var pi pathInfo
			pi.Path = path2
			pi.Type = pathInfoTypeDir
			pi.SavePath = filepath.Join(filepath.Base(path1), strings.TrimPrefix(path2, path1))
			if !info.IsDir() {
				pi.Type = pathInfoTypeFile
				pi.Size = info.Size()
				pi.SavePath = filepath.Dir(pi.SavePath)
			}
			respPaths = append(respPaths, pi)
			return nil
		})
		if err != nil {
			logrus.Error("walk file error: ", err)
			respCommonError(conn, err.Error())
			return err
		}
	}
	body, err := json.Marshal(respPaths)
	if err != nil {
		logrus.Error("json marshal failed, err:", err)
		respCommonError(conn, err.Error())
		return err
	}
	return sendMsgWithBody(conn, "", DataTypeFilePaths, body)
}

func sendImage(conn net.Conn) {
	imageName := time.Now().Format("20060102150405") + ".png"
	_ = sendMsgWithBody(conn, imageName, DataTypeClipImage, clipboardWatchData)
}

func sendText(conn net.Conn) {
	_ = sendMsgWithBody(conn, "", DataTypeText, clipboardWatchData)
}

func respCommonError(conn net.Conn, msg string) (ok bool) {
	return respError(conn, 400, msg)
}

func respError(conn net.Conn, code int, msg string) (ok bool) {
	var resp RespHead
	resp.Code = code
	resp.Msg = msg
	respBuf, err := json.Marshal(resp)
	if err != nil {
		logrus.Error("json marshal failed, err:", err)
		return false
	}
	var headLen = len(respBuf)
	var headLenBuf [4]byte
	binary.LittleEndian.PutUint32(headLenBuf[:], uint32(headLen))
	if _, err := conn.Write(headLenBuf[:]); err != nil {
		logrus.Error("write head len failed, err:", err)
		return false
	}
	if _, err := conn.Write(respBuf); err != nil {
		logrus.Error("write head failed, err:", err)
		return false
	}
	return true
}

func sendMsg(conn net.Conn, msg string) error {
	var resp RespHead
	resp.Code = 200
	resp.Msg = msg
	return sendHead(conn, resp)
}

func sendHead(conn net.Conn, head RespHead) error {
	respBuf, err := json.Marshal(head)
	if err != nil {
		logrus.Error("json marshal failed, err:", err)
		return err
	}
	// logrus.Debugln("respHead:", string(respBuf))
	logrus.Debugln("respHead:", head)
	var headLen = len(respBuf)

	var headLenBuf [4]byte
	binary.LittleEndian.PutUint32(headLenBuf[:], uint32(headLen))
	if _, err := conn.Write(headLenBuf[:]); err != nil {
		logrus.Error("write head len failed, err:", err)
		return err
	}
	if _, err := conn.Write(respBuf); err != nil {
		logrus.Error("write head failed, err:", err)
		return err
	}
	return nil
}

func sendMsgWithBody(conn net.Conn, msg string, datatype string, body []byte) error {
	var resp RespHead
	resp.Code = 200
	resp.Msg = msg
	resp.DataType = datatype
	resp.DataLen = int64(len(body))
	respBuf, err := json.Marshal(resp)
	if err != nil {
		logrus.Error("json marshal failed, err:", err)
		return err
	}
	var headLen = len(respBuf)
	// fmt.Println("headLen:", headLen, "head:", string(respBuf))
	var headLenBuf [4]byte
	binary.LittleEndian.PutUint32(headLenBuf[:], uint32(headLen))
	if _, err := conn.Write(headLenBuf[:]); err != nil {
		logrus.Error("write head len failed, err:", err)
		return err
	}
	if _, err := conn.Write(respBuf); err != nil {
		logrus.Error("write head failed, err:", err)
		return err
	}
	if _, err := conn.Write(body); err != nil {
		logrus.Error("write body failed, err:", err)
		return err
	}
	return nil
}

File: windSend-rs/src/route/route.rs
use serde::{Deserialize, Serialize};
use tokio::io::AsyncReadExt;
use tokio::net::TcpStream;
use tokio_rustls::server::TlsStream;
use tracing::{debug, error, trace};

#[derive(Debug, Serialize, Deserialize, Default)]
pub enum RouteAction {
    #[default]
    Unknown,
    #[serde(rename = "ping")]
    Ping,
    #[serde(rename = "pasteText")]
    PasteText,
    #[serde(rename = "pasteFile")]
    PasteFile,
    #[serde(rename = "copy")]
    Copy,
    #[serde(rename = "download")]
    Download,
}

#[derive(Debug, Serialize, Deserialize, Default)]
pub struct RouteHead {
    pub action: RouteAction,
    #[serde(rename = "timeIp")]
    pub time_ip: String,
    #[serde(rename = "fileID")]
    pub file_id: u32,
    #[serde(rename = "fileSize")]
    pub file_size: i64,
    #[serde(rename = "path")]
    pub down_path: String,
    /// 上传文件的名称
    #[serde(rename = "name")]
    pub file_name: String,
    pub start: i64,
    pub end: i64,
    #[serde(rename = "dataLen")]
    pub data_len: i64,
    /// 操作ID
    #[serde(rename = "opID")]
    pub op_id: u32,
    /// 此次操作想要上传的文件数量
    #[serde(rename = "filesCountInThisOp")]
    pub files_count_in_this_op: i32,
}

#[derive(Debug, Serialize)]
pub struct RouteRespHead<'a> {
    pub code: i32,
    pub msg: &'a String,
    /// 客户端copy时返回的数据类型(text, image, file)
    #[serde(rename = "dataType")]
    pub data_type: RouteDataType,
    /// 如果body有数据，返回数据的长度
    #[serde(rename = "dataLen")]
    pub data_len: i64,
    pub paths: Vec<RoutePathInfo>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct RoutePathInfo {
    pub path: String,
    pub size: u64,
}

#[derive(Debug, Serialize, Deserialize)]
pub enum RouteDataType {
    #[serde(rename = "text")]
    Text,
    #[serde(rename = "clip-image")]
    ClipImage,
    #[serde(rename = "files")]
    Files,
    #[serde(rename = "binary")]
    Binary,
}
static TIME_FORMAT: &str = "%Y-%m-%d %H:%M:%S";
static EXAMINE_TIME_STR: &str = "2023-10-10 01:45:32";
static MAX_TIME_DIFF: i64 = 10;

pub async fn main_process(mut conn: tokio_rustls::server::TlsStream<tokio::net::TcpStream>) {
    let head = common_auth(&mut conn).await;
    if head.is_err() {
        return;
    }
    let head = head.unwrap();
    debug!("head: {:?}", head);
    match head.action {
        RouteAction::Ping => {
            crate::route::paste::ping_handler(&mut conn, head)
                .await
                .ok();
        }
        RouteAction::PasteText => {
            crate::route::paste::paste_text_handler(&mut conn, head).await;
        }
        RouteAction::PasteFile => {
            crate::route::paste::paste_file_handler(&mut conn, head).await;
        }
        RouteAction::Copy => {
            crate::route::copy::copy_handler(&mut conn).await;
        }
        RouteAction::Download => {
            crate::route::copy::download_handler(&mut conn, head).await;
        }
        _ => {
            crate::route::resp::resp_error_msg(
                &mut conn,
                &format!("unknown action: {:?}", head.action),
            )
            .await
            .ok();
            error!("unknown action: {:?}", head.action);
        }
    }
    use tokio::io::AsyncWriteExt;
    conn.flush()
        .await
        .map_err(|e| error!("flush failed, err: {}", e))
        .ok();
}

pub async fn common_auth(conn: &mut TlsStream<TcpStream>) -> Result<RouteHead, ()> {
    let head_buf_size = 1024;
    let mut head_buf = vec![0u8; head_buf_size];
    // 读取json长度
    let mut head_len = [0u8; 4];
    conn.read_exact(&mut head_len)
        .await
        .map_err(|e| error!("read head len failed, err: {}", e))
        .ok();
    let head_len = i32::from_le_bytes(head_len);
    if head_len > head_buf_size as i32 || head_len <= 0 {
        error!("invalid head len: {}", head_len);
        return Err(());
    }
    trace!("head_len: {}", head_len);
    head_buf.resize(head_len as usize, 0);
    // 读取json
    conn.read_exact(&mut head_buf)
        .await
        .map_err(|e| error!("read head failed, err: {}", e))
        .ok();
    let head: RouteHead = serde_json::from_slice(&head_buf)
        .map_err(|e| error!("json unmarshal failed, err: {}", e))?;
    if head.time_ip.is_empty() {
        error!("time-ip is empty");
        return Err(());
    }
    debug!("head: {:?}", head);

    let time_and_ip_bytes =
        hex::decode(&head.time_ip).map_err(|e| error!("hex decode failed, err: {}", e))?;
    let decrypted = crate::config::get_cryptor()
        .map_err(|e| error!("get_cryptor failed, err: {}", e))?
        .decrypt(&time_and_ip_bytes);
    // .map_err(|e| error!("decrypt failed, err: {}", e))?;
    if let Err(e) = decrypted {
        let remote_ip = conn
            .get_ref()
            .0
            .peer_addr()
            .map_err(|e| error!("get peer addr failed, err: {}", e));
        match remote_ip {
            Ok(remote_ip) => {
                error!(
                    "decrypt failed, err: {}, remote_ip: {}",
                    e,
                    remote_ip.ip().to_string()
                );
            }
            Err(_) => {
                error!("get peer addr failed");
                error!("decrypt failed, err: {}", e)
            }
        }
        return Err(());
    }
    let decrypted = decrypted.unwrap();

    let time_and_ip_str =
        String::from_utf8(decrypted).map_err(|e| error!("utf8 decode failed, err: {}", e))?;
    let time_len = EXAMINE_TIME_STR.len();
    if time_and_ip_str.len() < time_len {
        error!("time-ip is too short");
        return Err(());
    }
    let time_str = &time_and_ip_str[..time_len];
    let ip = &time_and_ip_str[time_len + 1..];
    trace!("time_str: {}, ip: {}", time_str, ip);
    let t = chrono::NaiveDateTime::parse_from_str(time_str, TIME_FORMAT)
        .map_err(|e| error!("parse time failed, err: {}", e))?;
    if chrono::Utc::now()
        .signed_duration_since(t.and_utc())
        .num_seconds()
        > MAX_TIME_DIFF
    {
        error!("time expired: {}", t);
        return Err(());
    }
    let myipv4 = conn
        .get_ref()
        .0
        .local_addr()
        .map_err(|e| error!("get local addr failed, err: {}", e))?;
    trace!("ip: {}, myipv4: {}", ip, myipv4.ip());
    if ip != myipv4.ip().to_string() {
        error!("ip not match: {} != {}", ip, myipv4.ip());
        return Err(());
    }
    Ok(head)
}
